File: lib/builder.rb
require 'ast/sequence'
require 'ast/variable'
require 'ast/word'

class Builder
  def initialize(scope = Object.new)
    self.scope = scope
  end

  def build_sequence(expressions)
    AST::Sequence.new(expressions)
  end

  def build_variable(name)
    AST::Variable.new(name, scope)
  end

  def build_word(name)
    AST::Word.new(name)
  end

  private

  attr_accessor :scope
end

File: src/builder.js
import AST from './ast';
import { List } from 'immutable';
import Scope from './scope';

export default class {
  constructor(scope = new Scope()) {
    this.scope = scope;
  }

  buildKeyword(name) {
    return new AST.Keyword({ name });
  }

  buildSequence(expressions) {
    return new AST.Sequence({ expressions: List(expressions) });
  }

  buildVariable(name) {
    return new AST.Variable({ name, scope: this.scope });
  }
};
